src/t_hash.c

```diff
@@ -406,22 +423,41 @@ static uint64_t listpackExGetMinExpire(robj *o) {
 }
 
 /* Walk over fields and delete the expired ones. */
-void listpackExExpire(robj *o, ExpireInfo *info) {
+void listpackExExpire(redisDb *db, robj *o, ExpireInfo *info) {
     serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_EX);
-    uint64_t min = EB_EXPIRE_TIME_INVALID;
+    uint64_t expired = 0, min = EB_EXPIRE_TIME_INVALID;
+    unsigned char *ptr;
     listpackEx *lpt = o->ptr;
 
-    struct lpFingArgs r = {
-        .max_to_search = info->maxToExpire,
-        .expire_time = info->now
-    };
+    ptr = lpFirst(lpt->lp);
 
-    lpFindCb(lpt->lp, NULL, &r, cbFindInListpack, 0);
-    info->itemsExpired += r.expired;
+    while (ptr != NULL && (info->itemsExpired < info->maxToExpire)) {
+        long long val;
+        int64_t flen;
+        unsigned char intbuf[LP_INTBUF_SIZE], *fref;
+
+        fref = lpGet(ptr, &flen, intbuf);
+
+        ptr = lpNext(lpt->lp, ptr);
+        serverAssert(ptr);
+        ptr = lpNext(lpt->lp, ptr);
+        serverAssert(ptr && lpGetIntegerValue(ptr, &val));
+
+        /* Fields are ordered by expiry time. If we reached to a non-expired
+         * or a non-volatile field, we know rest is not yet expired. */
+        if (val == HASH_LP_NO_TTL || (uint64_t) val > info->now)
+            break;
+
+        propagateHashFieldDeletion(db, ((listpackEx *) o->ptr)->key, (char *)((fref) ? fref : intbuf), flen);
 
-    /* Delete all the expired fields in one go */
-    if (r.expired > 0)
-        lpt->lp = lpDeleteRange(lpt->lp, 0, r.expired * 3);
+        ptr = lpNext(lpt->lp, ptr);
```

Comment:
Just a note to all of us, maybe we can implement this logic with lpFindCb() in future to avoid performance cost of lpNext() calls.